R/Previewer.R

```diff
@@ -104,378 +115,232 @@ reporter_previewer_srv <- function(id,
-          class = if (nr_cards) "" else "disabled"
-        ),
-        load = shiny::actionButton(
-          ns("load_reporter_previewer"),
-          class = "teal-reporter simple_report_button",
-          `data-val` = shiny::restoreInput(id = ns("load_reporter_previewer"), default = NULL),
-          shiny::tags$span(
-            "Load Report", shiny::icon("upload")
-          )
-        ),
-        reset = reset_report_button_ui(ns("resetButtonPreviewer"), label = "Reset Report")
-      )
+    download_report_button_srv(
+      "download",
+      reporter = reporter,
+      global_knitr = global_knitr,
+      rmd_output = rmd_output,
+      rmd_yaml_args = rmd_yaml_args
+    )
+    report_load_srv("load", reporter = reporter)
+    reset_report_button_srv("reset", reporter = reporter)
 
-      shiny::tags$div(
-        id = "previewer_reporter_encoding",
-        shiny::tags$h3("Download the Report"),
-        shiny::tags$hr(),
-        reporter_download_inputs(
-          rmd_yaml_args = rmd_yaml_args,
-          rmd_output = rmd_output,
-          showrcode = any_rcode_block(reporter),
-          session = session
-        ),
-        shiny::tags$div(
-          id = "previewer_reporter_buttons",
-          class = "previewer_buttons_line",
-          lapply(previewer_buttons_list[previewer_buttons], shiny::tags$div)
-        )
-      )
+    current_cards <- shiny::reactiveVal()
+    insert_cards <- shiny::reactiveVal()
+    remove_cards <- shiny::reactiveVal()
+    shiny::observeEvent(reporter$get_reactive_add_card(), {
+      to_add <- reporter$get_cards()[!reporter$get_cards() %in% current_cards()] # because setdiff loses names
+      to_remove <- current_cards()[!current_cards() %in% reporter$get_cards()]
+      if (length(to_add)) insert_cards(to_add)
+      if (length(to_remove)) remove_cards(to_remove)
+      current_cards(reporter$get_cards())
     })
 
-    output$pcards <- shiny::renderUI({
-      reporter$get_reactive_add_card()
-      input$card_remove_id
-      input$card_down_id
-      input$card_up_id
-
-      cards <- reporter$get_cards()
-
-      if (length(cards)) {
-        shiny::tags$div(
-          class = "panel-group accordion",
-          id = "reporter_previewer_panel",
-          lapply(seq_along(cards), function(ic) {
-            previewer_collapse_item(ic, cards[[ic]]$get_name(), cards[[ic]]$get_content())
-          })
+    shiny::observeEvent(insert_cards(), {
+      cards <- insert_cards()
+      lapply(names(cards), function(card_name) {
+        bslib::accordion_panel_insert(
+          id = "reporter_cards",
+          reporter_previewer_card_ui(id = session$ns(card_name), card_name = card_name)
         )
-      } else {
-        shiny::tags$div(
-          id = "reporter_previewer_panel_no_cards",
-          shiny::tags$p(
-            class = "text-danger mt-4",
-            shiny::tags$strong("No Cards added")
-          )
+        reporter_previewer_card_srv(
+          id = card_name,
```

Comment:
This is a bad idea to use the name verbatim here as it will create a bad DOM

![image](https://github.com/user-attachments/assets/4ed0852b-24f6-4194-8af7-1c5f25292c9c)

I would argue that `Reporter$append_cards` should take the name and store it as an attribute and generate a random string as id _(using `rlang::hash` with the name)_

Reply:
fixed in #325